### PR TITLE
PWX-35477 : Support Openshift Prometheus for portworx monitoring on OCP 4.14

### PR DESCRIPTION
**Pull request by nikita-bhatia**
Issue : OCP upgrade from 4.13.x to 4.14.5 is blocked due to PX prometheus. From OCP 4.14, PX Prometheus stopped working.

Resolution : From OCP 4.14, Portworx deployment on OCP should should use Openshift prometheus stack for monitoring Portworx metrics instead of deploying PX-Prometheus. We need to expose Portworx metrics to Openshift.
To do this, user-workload-monitoring has to be enabled on OCP cluster to use openshift's monitoring and alerting for portworx.

**What this PR does / why we need it**:
1. Default autopilot URL changed from  http://px-prometheus:9090/ to https://$THANOS_QUERIER_HOST 
2. Creating autopilot-auth-token-secret  to store ocp token and ca.cert
3. Mount ocp token and ca.cert as file in autopilot pod with file names ca-certificates.crt and token. 

**Which issue(s) this PR fixes** (optional)
Closes # https://portworx.atlassian.net/browse/PWX-35477

**Special notes for your reviewer**:
TBD

**Testing done**
TBD

### CODE REVIEW

**Comment by zoxpx** [comment]
@piyush-nimbalkar -- would you mind reviewing this?

It is a fairly complicated PR, and you were the author of the original `drivers/storage/portworx/component/autopilot.go` where the bulk of the changes are from.

File: drivers/storage/portworx/component/autopilot.go [hunk 1]
```diff
@@ -659,6 +754,125 @@ func (c *autopilot) getDesiredVolumesAndMounts(
...
+			}
+
+			// Retrieve the ca.cert data from the secret as []byte
+			cert, ok := secret.Data["ca.crt"]
+			if !ok {
+				return encodedToken, caCert, fmt.Errorf("cert not found in secret")
+			}
+
+			encodedToken = string(tokenBytes)
+			caCert = string(cert)
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/autopilot.go:788]: break here as we have already found the secret

File: drivers/storage/portworx/component/autopilot.go [hunk 2]
```diff
@@ -131,6 +182,12 @@ func (c *autopilot) Reconcile(cluster *corev1.StorageCluster) error {
 	if err := c.createDeployment(cluster, ownerRef); err != nil {
 		return err
 	}
+
+	if IsOCPUserWorkloadSupported(c.k8sClient, c) {
+		if err := c.createSecret(cluster.Namespace, ownerRef); err != nil {
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/autopilot.go:187]: Shouldn't we create the secret before we create the deployment? That way the chances of the AP pod failing are lower?

File: drivers/storage/portworx/component/autopilot.go [hunk 3]
```diff
@@ -659,6 +754,125 @@ func (c *autopilot) getDesiredVolumesAndMounts(
...
+				return encodedToken, caCert, fmt.Errorf("cert not found in secret")
+			}
+
+			encodedToken = string(tokenBytes)
+			caCert = string(cert)
+		}
+	}
+
+	if !secretFound {
+		return "", "", fmt.Errorf("prometheus-user-workload-token not found. Please make sure that user workload is enabled in openshift")
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/autopilot.go:793]: ... make sure that user workload **monitoring** is enabled ...

File: drivers/storage/portworx/util/util.go [hunk 4]
```diff
@@ -1770,3 +1780,80 @@ func AppendUserVolumeMounts(
...
+			context.TODO(),
+			types.NamespacedName{
+				Name: OpenshiftAPIServer,
+			},
+			operator,
+		)
+
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, err
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/util/util.go:1807]: We should not return error if the object is not found. It is an indicator that this is not a supported openshift environment.

File: drivers/storage/portworx/component/autopilot.go [hunk 5]
```diff
@@ -45,7 +47,12 @@ const (
 	AutopilotDefaultProviderEndpoint = "http://px-prometheus:9090"
 	// AutopilotDefaultReviewersKey is a key for default reviewers array in gitops config map
 	AutopilotDefaultReviewersKey = "defaultReviewers"
-	defaultAutopilotCPU          = "0.1"
+	// OCPPrometheusUserWorkloadSecretPrefix name of OCP user-workload Prometheus secret
+	OCPPrometheusUserWorkloadSecretPrefix = "prometheus-user-workload-token"
+	// Autopilot Secret name for prometheus-user-workload-token
+	AutopilotSecretName                 = "autopilot-prometheus-auth"
+	defaultAutopilotCPU                 = "0.1"
+	OpenshiftPrometheusSupportedVersion = "4.14"
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/component/autopilot.go:55]: Can we also move this constant to pxutil and use it in portworx.go and the tests instead of hard coding 4.14 there?

**Review by piyush-nimbalkar — APPROVED** [review]
LGTM after addressing the comments.

@nikita-bhatia can we add more UTs for all corner cases like if we don't find the ocp secret or if the right data is not present in it? What happens when we disable and enable autopilot again - it should have the same deployment and the secret should have been deleted when autopilot was disabled.

Let's merge this PR after addressing comments and then add just the UTs in a different PR.
